packages/webdriver-image-comparison/src/helpers/utils.ts

```diff
@@ -325,3 +331,161 @@ export function isStorybook(){
+        </html>`
+
+    const base64Html = Buffer.from(htmlContent).toString('base64')
+
+    await url(`data:text/html;base64,${base64Html}`)
+
+    if (isIOS) {
+        await executor(checkMetaTag)
+    }
+}
+
+/**
+ * Execute a native click
+ */
+export async function executeNativeClick({ executor, isIOS, x, y }:{executor: Executor, isIOS:boolean, x: number, y: number}): Promise<void> {
+    if (isIOS) {
+        return executor('mobile: tap', { x, y })
+    }
+
+    try {
+        // The `clickGesture` is not working on Appium 1, only on Appium 2
+        await executor('mobile: clickGesture', { x, y })
+    } catch (error: unknown) {
+        if (
+            error instanceof Error &&
+          /WebDriverError: Unknown mobile command.*?(clickGesture|tap)/i.test(error.message)
+        ) {
+            console.log(
+                'Error executing `clickGesture`, falling back to `doubleClickGesture`. This likely means you are using Appium 1. Is this intentional?'
+            )
+            await executor('mobile: doubleClickGesture', { x, y })
+        } else {
+            throw error
+        }
+    }
+}
+
+/**
+ * Get the mobile viewport position, we determine this by:
+ * 1. Loading a base64 HTML page
+ * 2. Injecting an overlay on top of the webview with an event listener that stores the click position in the webview
+ * 3. Clicking on the overlay in the center of the screen with a native click
+ * 4. Getting the data from the overlay and removing it
+ * 5. Calculating the position of the viewport based on the click position of the native click vs the overlay
+ * 6. Returning the calculated values
+ */
+export async function getMobileViewPortPosition({
+    initialDeviceRectangles,
+    isAndroid,
+    isIOS,
+    isNativeContext,
+    methods: {
+        executor,
+        getUrl,
+        url,
+    },
+    nativeWebScreenshot,
+    screenHeight,
+    screenWidth,
+}: GetMobileViewPortPositionOptions): Promise<DeviceRectangles> {
+
+    if (!isNativeContext && (isIOS || (isAndroid && nativeWebScreenshot))) {
+        const currentUrl = await getUrl()
+        // 1. Load a base64 HTML page
+        await loadBase64Html({ executor, isIOS, url })
+        // 2. Inject an overlay on top of the webview with an event listener that stores the click position in the webview
+        await executor(injectWebviewOverlay, isAndroid)
+        // 3. Click on the overlay in the center of the screen with a native click
+        const nativeClickX = screenWidth / 2
+        const nativeClickY = screenHeight / 2
+        await executeNativeClick({ executor, isIOS, x: nativeClickX, y: nativeClickY })
+        // We need to wait a bit here, otherwise the click is not registered
+        await waitFor(100)
+        // 4a. Get the data from the overlay and remove it
+        const { y, x, width, height } = await executor(getMobileWebviewClickAndDimensions, '[data-test="ics-overlay"]')
+        // 4.b reset the url
+        await url(currentUrl)
```

Comment:
Can you explain why we have to do things like this?

Reply:
A screenshot on a device, especially on iOS, is a screenshot of the full screen. Meaning including the header with the time, wifi, and the bottom bar. The viewport with the actual data, is between that. Appium won't give this data back.
The position of the webview will differ per device. With this method we do an "exact" calculation of the position of the webview so we can:

- better block out the stuff we don't need during the comparison
- cut out elements from the device screenshot
- create better screenshots when we stitch screenshots for the fullpage screenshot method